Add rule for coordination between services of different contexts

mLeveIST opened the pull request:
## Overview
This change provides a way to represent coordination between services of different contexts, with the addition of a Coordination rule in the Application and Process Layer of Bounded Contexts. Example in CML:
```
ContextMap TestMap {
    contains ContextA
    contains ContextB

    ContextA <- ContextB
}

BoundedContext ContextA {
    Application {
        Coordination CoordinationA {
            ContextA::ServiceA::stepA;
            ContextB::ServiceB::stepB;
        }

        Service ServiceA {
            void stepA();
        }
    }
}

BoundedContext ContextB {
    Application {
        Service ServiceB {
            void stepB();
        }
    }
}
```
A Coordination can be visualised in a BPMN diagram with Sketch Miner, where each context will be represented as a different actor, and each operation as a task. The example above generates the following code:
```
ContextA: service ServiceA.stepA
ContextB: service ServiceB.stepB
```

## Motivation

The main drive for this rule is to be able to model in CML processes that span multiple Bounded Contexts, like Sagas. CML does already provides a Flow rule to model processes, but these processes are restricted to a single Bounded Context and the rule has a focus on modelling Event/Command flows for Event Storming.

## Technical Details

The following sections specify details on the implementation of the rule.

### Validation

Since the steps of a coordination are composed of references to other rules, the following verifications are done to avoid ambiguity when pinpointing the step operation:
- A reference to an outer context must be reachable by Context Map relationships
- A reference to a service must originate from an Application Layer Service
- A reference to an operation should be unique within the service

### IDE integration

The following features were implemented for the IDE when writing in CML:
- Formatting for the rule
- Step wise visualisation of validation errors
- BPMN Sketch Miner diagram code generation
- Quick-fix to generate and open a Coordination in BPMN Sketch Miner website

stefan-ka commented:
Thanks for creating the PR @mLeveIST ;) 
I will probably need a few days until you get a response (review comments)...
Best regards,
Stefan

stefan-ka commented on org.contextmapper.dsl/src/org/contextmapper/dsl/validation/ApplicationCoordinationSemanticsValidator.java:
```diff
@@ -0,0 +1,107 @@
+	}
+	
+	@Check
+	public void stepComponentsMustBeCorrectlyReferenced(final CoordinationStep coordinationStep) {
```
Thats a very long method with multiple checks. Could you maybe split it into multiple methods? That blocks that are marked with comments could be separate methods I guess.

mLeveIST commented:
Hello Stefan,

Thank you for the review!
I will tag you once I have addressed the comments :)

Best regards,
Miguel

mLeveIST commented on org.contextmapper.dsl.tests/src/org/contextmapper/dsl/generators/sketchminer/SketchMinerLinkCreatorTest.java:
```diff
@@ -95,6 +96,38 @@ public void canCreateSketchMinerLinkWithEObjectWin() throws IOException {
+	
+	@Test
+	@DisabledOnOs(OS.WINDOWS)
+	public void canCreateSketchMinerLinkForCoordinationModelUnix() throws IOException {
```
Unfortunately I could not test it on a Windows :(

mLeveIST commented:
Hello @stefan-ka,

I added most of the changes you requested, just missing the test for the Windows machine...
I also added some additional tests for the code coverage.

Regarding the documentation, can this be added to the Application and Process Layer page?

If you have the time this or next week, I wouldn't mind having a small meeting to discuss the results of Coordinations as is, and if you think it is worthwhile to expand upon it in the future (having the type of coordination, the synchronicity of the steps, if they can relate to flows).

Best regards,
Miguel

stefan-ka commented:
Hi @mLeveIST 

Thank you! I will try to do the tests on Windows... I will come back to you regarding a meeting. Probably next week earliest. 

Best regards,
Stefan

stefan-ka commented:
Sorry @mLeveIST, I forgot to answer your question regarding documentation.

I would say, yes. Or create a new page but then we have to link it on the application layer page. 

Best regards,
Stefan

stefan-ka commented:
Hi @mLeveIST 

I just realized that we have many other issues with unit tests on Windows 🫣 
I suggest we fix that we a separate PR later. I have to fix several other tests to run on Windows...

From my point of view we can then merge it. But lets give @socadk a chance to test the new feature as well.

We'll come back to you.

Best regards,
Stefan

mLeveIST commented:
Hi @stefan-ka,

Thank you for the update! That's great to know.
Regarding documentation, I'll have a pull request for it and also the examples repo soon, most likely sometime next week if that's ok.

Best regards,
Miguel

stefan-ka commented:
Perfect, thanks @mLeveIST!
